module/documents/actor/trait.mjs

```diff
@@ -84,29 +133,93 @@ export async function choices(trait, chosen=new Set()) {
       let type = foundry.utils.getProperty(index, keyPath);
       if ( map?.[type] ) type = map[type];
 
-      const entry = { label: index.name, chosen: chosen.has(key) };
-
       // No category for this type, add at top level
-      if ( !data[type] ) data[key] = entry;
+      if ( !config[type] ) config[key] = { label: index.name };
 
-      // Add as child to appropriate category
+      // Add as child of appropriate category
       else {
-        data[type].children ??= {};
-        data[type].children[key] = entry;
+        config[type].children ??= {};
+        config[type].children[key] = { label: index.name };
       }
     }
   }
 
-  // Sort Categories
-  if ( traitConfig.sortCategories ) data = dnd5e.utils.sortObjectEntries(data, "label");
+  return config;
+}
+
+/* -------------------------------------------- */
 
-  // Sort Children
-  for ( const category of Object.values(data) ) {
-    if ( !category.children ) continue;
-    category.children = dnd5e.utils.sortObjectEntries(category.children, "label");
+/**
+ * Get a list of choices for a specific trait.
+ * @param {string} trait                      Trait as defined in `CONFIG.DND5E.traits`.
+ * @param {object} [options={}]
+ * @param {Set<string>} [options.chosen=[]]   Optional list of keys to be marked as chosen.
+ * @param {boolean} [options.prefixed=false]  Should keys be prefixed with trait type?
+ * @param {boolean} [options.any=false]       Should the "Any" option be added to each category?
+ * @returns {SelectChoices}                   Object mapping proficiency ids to choice objects.
+ */
+export async function choices(trait, { chosen=new Set(), prefixed=false, any=false }={}) {
+  const traitConfig = CONFIG.DND5E.traits[trait];
+  if ( !traitConfig ) return new SelectChoices();
+  if ( foundry.utils.getType(chosen) === "Array" ) chosen = new Set(chosen);
+  const categoryData = await categories(trait);
+
+  let result = {};
+  if ( prefixed && any ) {
+    const key = `${trait}:*`;
+    result[key] = {
+      label: keyLabel(key).titleCase(),
+      chosen: chosen.has(key), sorting: false, wildcard: true
+    };
   }
 
-  return data;
+  const prepareCategory = (key, data, result, prefix) => {
+    let label = _innerLabel(data, traitConfig);
+    if ( !label ) label = key;
+    if ( prefixed ) key = `${prefix}:${key}`;
+    result[key] = {
+      label: game.i18n.localize(label),
+      chosen: chosen.has(key),
+      sorting: traitConfig.sortCategories === true
+    };
+    if ( data.children ) {
+      const children = result[key].children = {};
+      if ( prefixed && any ) {
+        const anyKey = `${key}:*`;
+        children[anyKey] = {
+          label: keyLabel(anyKey).titleCase(),
+          chosen: chosen.has(anyKey), sorting: false, wildcard: true
+        };
+      }
+      Object.entries(data.children).forEach(([k, v]) => prepareCategory(k, v, children, key));
+    }
+  };
```

Comment:
This is perhaps a silly suggestion, but could this be simplified somewhat by delegating the prefixing to the `SelectChoices` constructor (maybe with an option)? Things started falling into place a lot better once I realised that the children are already prefixed when supplied to the `SelectChoices` constructor. I thought we should maybe document that then, but then I thought we might be able to actually do the prefixing in the constructor, and then it would be obvious.

Maybe that's not feasible, I'm not sure, but maybe you could try it out. Otherwise, ultimately, this is just utility code so it's fine as-is as well with just a small mention in the constructor JSDoc that children are supplied already-prefixed.

Reply:
I think the issue is that generating the `chosen` value for each choice requires the fully prefixed key, so if we left prefixing to the `SelectChoices` object we'd probably have to iterate through the entire structure again to set `chosen` or generate the keys twice.

Reply:
Ah ok, I see. Can we just document the `SelectChoicesEntry` typedef then to mention that the `children` are required to be already-prefixed? I suppose we could add an option to perform the prefixing in `SelectChoices` if we ever needed it. I guess this method is the only place we instantiate a `SelectChoices` object though currently.

Reply:
I'll add a note about providing prefixed keys if desired, though we do also use this method with un-prefixed keys in `TraitSelector` so it isn't a requirement.